docs/travis.md

```diff
@@ -0,0 +1,102 @@
+# Travis CI
+
+The Zulip server uses [Travis CI](https://travis-ci.org/) for its
+continuous integration.  This page documents useful tools and tips to
+know about when using Travis CI and debugging issues with it.
+
+## Goals
+
+The overall goal of our Travis CI setup is to catch as many possible
+future bugs in Zulip as possible, minimizing delay and false positives
+that waste the time of developers around the Zulip community
+debugging.  There are a few implications of this overall goal:
+
+* If a test is failing nondeterministically in Travis CI, we consider
+  that to be an urgent problem.
+* If the tests become a lot slower, that is also an urgent problem.
+* Everything we do in CI should be to the extent possible runnable in
+  an environment where one can iterate quickly, so that developers
+  only need to wait 10 minutes for Travis CI to run repeatedly when
+  modifying the Travis CI configuration itself.
+
+## Configuration
+
+The main Travis configuration file is `.travis.yml`.  The specific
+test suites we have are listed in the `matrix` section in that file.
+We've configured it to use a few helper scripts for each job:
+
+* `tools/travis/setup-$TEST_SUITE`: A Zulip script that sets up the
+  test environment (E.g. installing dependencies).
+  * For the backend and frontend suites, this is a thin wrapper around
+    `tools/provision`, aka the development environment provision script.
+  * For the production suite, this is a more complicated process
+    because of all the packages Travis installs.  See the comments in
+    `tools/travis/setup-production` for details.
+* `tools/travis/$TEST_SUITE`: A zulip script containing the actual
+  code for the test.
+
+The main purpose of the distinction between the two is that if the
+`setup-backend` job fails, Travis CI will report it as a build
+infrastructure failure, whereas if the `backend` job fails, it'll be
+reported as a test failure.
+
+An important detail is that Travis CI hides most build phases other
+than the actual test; you can see this easily by looking at the line
+numbers in the Travis CI output.  There are actually a bunch of phases
+(e.g. the project's setup job, downloading caches near the beginning,
+uploading caches at the end, etc.), and if you're debugging our
+configuration, you'll want to look at these closely.
+
+## Useful debugging tips and tools
+
+* For performance issues,
+  [this travis stats tool](https://scribu.github.io/travis-stats/#zulip/zulip/master)
+  can give you test runtime history data that can help with
+  determining when a performance issue was introduced and whether it
+  was fixed.
+
+* You can [sign up your personal repo for Travis CI][travis-fork] so
+  that every remote branch you push will be tested, which can be
+  helpful when debugging something complicated.
+
+[travis-fork]: git-guide.html#step-3-configure-travis-ci-continuous-integration
+
```

Comment:
Shouldn't we mention about using `ts` to time tests here. I would add something like this after first point about performance issues.
* Also one can always use `ts` to profile the tests and figure out what runs
  slower. Checkout [`da731c`](https://github.com/zulip/zulip/commit/da731c)
  this commit to see how such a profiling could be done.